src/ray/observability/tests/ray_event_recorder_test.cc

```diff
@@ -0,0 +1,122 @@
+// Copyright 2025 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/observability/ray_event_recorder.h"
+
+#include <chrono>
+#include <mutex>
+#include <thread>
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/ray_config.h"
+#include "ray/observability/ray_driver_job_definition_event.h"
+#include "ray/observability/ray_driver_job_execution_event.h"
+#include "src/ray/protobuf/gcs.pb.h"
+#include "src/ray/protobuf/public/events_base_event.pb.h"
+#include "src/ray/protobuf/public/events_driver_job_execution_event.pb.h"
+
+namespace ray {
+namespace observability {
+
+class FakeEventAggregatorClient : public rpc::EventAggregatorClient {
+ public:
+  FakeEventAggregatorClient(std::vector<rpc::events::RayEvent> &recorded_events,
+                            std::mutex &mutex)
+      : recorded_events_(recorded_events), mutex_(mutex) {}
+
+  void AddEvents(
+      const rpc::events::AddEventsRequest &request,
+      const rpc::ClientCallback<rpc::events::AddEventsReply> &callback) override {
+    std::lock_guard<std::mutex> lock(mutex_);
+    for (const auto &event : request.events_data().events()) {
+      recorded_events_.push_back(event);
+    }
+    callback(Status::OK(), rpc::events::AddEventsReply{});
+  }
+
+ private:
+  std::vector<rpc::events::RayEvent> &recorded_events_;
+  std::mutex &mutex_;
+};
+
+class RayEventRecorderTest : public ::testing::Test {
+ public:
+  RayEventRecorderTest() {
+    fake_client_ = std::make_unique<FakeEventAggregatorClient>(recorded_events_, mutex_);
+    recorder_ = std::make_unique<RayEventRecorder>(*fake_client_, io_service_);
+    recorder_->StartExportingEvents();
+  }
+
+  instrumented_io_context io_service_;
+  std::vector<rpc::events::RayEvent> recorded_events_;
+  std::mutex mutex_;
```

Comment:
Don't need these two, they can just be private memebers of FakeEventAggregatorClient and we provide thread safe getter to access the data.